engine/src/test/java/io/camunda/zeebe/engine/processing/bpmn/compensation/CompensationEventExecutionTest.java

```diff
@@ -727,6 +718,145 @@ public void shouldDeleteAllSubscriptionWhenProcessIsCompletedWithoutTriggerCompe
+
+    ENGINE.deployment().withXmlResource(process).deploy();
+    final long processInstanceKey = ENGINE.processInstance().ofBpmnProcessId(PROCESS_ID).create();
+
+    ENGINE
+        .job()
+        .ofInstance(processInstanceKey)
+        .withType(SERVICE_TASK_TYPE_COMPENSABLE_ACTIVITY)
+        .complete();
+    ENGINE
+        .job()
+        .ofInstance(processInstanceKey)
+        .withType(SERVICE_TASK_TYPE_COMPENSABLE_ACTIVITY2)
+        .complete();
+    ENGINE
+        .job()
+        .ofInstance(processInstanceKey)
+        .withType(SERVICE_TASK_TYPE_COMPENSATION_HANDLER)
+        .complete();
+    ENGINE
+        .job()
+        .ofInstance(processInstanceKey)
+        .withType(SERVICE_TASK_TYPE_COMPENSATION_HANDLER2)
+        .complete();
+
+    // then
+    assertThat(
+            RecordingExporter.compensationSubscriptionRecords()
+                .withProcessInstanceKey(processInstanceKey)
+                .limit(8))
+        .extracting(
+            Record::getValueType,
+            Record::getIntent,
+            r -> r.getValue().getCompensableActivityScopeId())
+        .containsSubsequence(
+            tuple(
+                ValueType.COMPENSATION_SUBSCRIPTION,
+                CompensationSubscriptionIntent.TRIGGERED,
+                "embedded-subprocess"),
+            tuple(
+                ValueType.COMPENSATION_SUBSCRIPTION,
+                CompensationSubscriptionIntent.TRIGGERED,
+                "embedded-subprocess-2"));
+  }
+
+  @Test
+  public void shouldNotTriggerCompensationIfSubprocessIsNotCompleted() {
+    final var process =
+        createModelFromClasspathResource(
+            "/compensation/compensation-single-embedded-subprocess.bpmn");
+
+    ENGINE.deployment().withXmlResource(process).deploy();
+    final long processInstanceKey = ENGINE.processInstance().ofBpmnProcessId(PROCESS_ID).create();
+
+    ENGINE
+        .job()
+        .ofInstance(processInstanceKey)
+        .withType(SERVICE_TASK_TYPE_COMPENSABLE_ACTIVITY)
+        .complete();
+
+    ENGINE.job().ofInstance(processInstanceKey).withType("completableActivity").complete();
+
+    ENGINE.job().ofInstance(processInstanceKey).withType("NotActivableTask").complete();
+
+    // then
+    assertThat(
+            RecordingExporter.compensationSubscriptionRecords()
+                .withProcessInstanceKey(processInstanceKey)
+                .limit(4))
+        .extracting(
+            Record::getValueType, Record::getIntent, r -> r.getValue().getCompensableActivityId())
+        .containsSubsequence(
+            tuple(
+                ValueType.COMPENSATION_SUBSCRIPTION,
+                CompensationSubscriptionIntent.DELETED,
+                "ActivityToCompensate"),
+            tuple(
+                ValueType.COMPENSATION_SUBSCRIPTION,
+                CompensationSubscriptionIntent.DELETED,
+                "embedded-subprocess"));
+  }
+
+  @Test
+  public void shouldNotTriggerCompensationOnParentScope() {
```

Comment:
:x: The test doesn't verify that we don't trigger the compensation on the parent scope. 

The process should look similar to the following process. The compensation throw event is inside a subprocess and a compensation handler is outside the subprocess. 

![Screenshot from 2024-02-07 11-33-53](https://github.com/camunda/zeebe/assets/4305769/78fa89ff-ce75-476f-b634-6a57b139b800)

Given:
- Complete A

When:
- Complete B
- Complete Undo B

Then:
- The process instance is completed
- Undo A is not activated
- The compensation subscription of A is deleted